examples/rate_providers/rate_providers.py

```diff
@@ -0,0 +1,128 @@
+            yield offer
+
+
+async def prepare_activity(activity: Activity) -> Activity:
+    print(f"Prepare {activity}")
+    batch = await activity.execute_commands(
+        commands.Deploy(),
+        commands.Start(),
+        commands.SendFile("cubes.blend", "/golem/resource/scene.blend"),
+    )
+    await batch.wait(timeout=60)
+    return activity
+
+
+async def rate_provider(activity: Activity) -> str:
+    provider_id = activity.parent.parent.data.issuer_id
+    print(f"Rating {provider_id}")
+    frame_config = FRAME_CONFIG_TEMPLATE.copy()
+    fname = f"out{0:04d}.png"  # format used by `run-blender.sh` included in `PAYLOAD`
+    output_file = f"{provider_id}-{fname}"
+
+    start_time = datetime.utcnow()
+    batch = await activity.execute_commands(
+        commands.Run(f"echo '{json.dumps(frame_config)}' > /golem/work/params.json"),
+        commands.Run(["/golem/entrypoints/run-blender.sh"]),
+        commands.DownloadFile(f"/golem/output/{fname}", output_file),
+    )
+    await batch.wait(timeout=20)
+
+    end_time = datetime.utcnow()
+    scores[provider_id] = end_time - start_time
+
+    return output_file
+
+
+async def on_exception(func: Callable, args: Tuple, e: Exception) -> None:
+    activity = args[0]
+    print(f"Activity {activity} failed because of:\n{e}")
+    await activity.parent.close_all()
+
+
+async def main() -> None:
+    golem = GolemNode()
+    golem.event_bus.listen(DefaultLogger().on_event)
+
+    async with golem:
+        allocation = await golem.create_allocation(1.0)
+        payment_manager = DefaultPaymentManager(golem, allocation)
+        demand = await golem.create_demand(
+            PAYLOAD, allocations=[allocation], subnet=SUBNET
+        )
+
+        # `set_no_more_children` has to be called so `initial_proposals` will eventually stop yielding
+        asyncio.create_task(delay(timedelta(seconds=5), demand.set_no_more_children))
+
+        chain = Chain(
+            demand.initial_proposals(),
+            filter_rated_providers,
+            Map(default_negotiate),
+            Map(default_create_agreement),
+            Map(default_create_activity),
+            Map(prepare_activity),
+            Map(rate_provider, on_exception=on_exception),  # type: ignore[arg-type]
+            Buffer(10),
+            Limit(20),
+        )
+
+        async for result in chain:
+            print(f"Finished rating provider with {result}")
+
+        print(
+            f"TASK DONE:\n{[(s, scores[s]) for s in scores if scores[s] is not None]}"
+        )
+        await payment_manager.terminate_agreements()
+        await payment_manager.wait_for_invoices()
+
+
+if __name__ == "__main__":
+    loop = asyncio.get_event_loop()
+    task = loop.create_task(main())
+    try:
+        loop.run_until_complete(task)
+    except KeyboardInterrupt:
+        task.cancel()
+        try:
+            loop.run_until_complete(task)
+        except asyncio.CancelledError:
+            pass
```

Comment:
```suggestion
    asyncio.run(main())
```

Same as previous similar block.